<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dashboard for Influenza, SARS-COV-2 and RSV Assembly and Curation • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Dashboard for Influenza, SARS-COV-2 and RSV Assembly and Curation">
<meta property="og:description" content="MIRA is an interactive dashboard created using Dash, a python framework written on the top of Flask, Plotly.js and React.js. The Dash dashboard allows users to interactively create a metadata and config file for Influenza , SARS-COV-2 and RSV Assembly. Subsequently, it generates files necessary to upload via FTP to NCBI’s databases Genbank, BioSample, and SRA, as well as, GISAID.">
<meta property="og:image" content="https://github.com/CDCgov/MIRA/, https://cdcgov.github.io/MIRA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MIRA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MIRA in Docker Desktop

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/mira-dd-getting-started.html">Getting Started and Installation</a>
    </li>
    <li>
      <a href="articles/running-mira-dd-ont.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="articles/running-mira-dd-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="articles/upgrading-mira-dd.html">Upgrading MIRA Docker Desktop</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    MIRA-CLI

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/quick-start-ubuntu.html">Quick Start in Ubuntu</a>
    </li>
    <li>
      <a href="articles/getting-started.html">PC Getting Started</a>
    </li>
    <li>
      <a href="articles/mira-cli-mac-getting-started.html">MAC Getting Started</a>
    </li>
    <li>
      <a href="articles/mira-cli-installation.html">MIRA installation</a>
    </li>
    <li>
      <a href="articles/running-mira-cli-ont.html">Running Genome Assembly with MIRA-CLI -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="articles/running-mira-cli-illumina.html">Running Genome Assembly with MIRA-CLI -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="articles/upgrading-mira-cli.html">Upgrading MIRA-CLI</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lab Protocols

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Sequencing Template Generation</li>
    <li>
      <a href="articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a>
    </li>
    <li>
      <a href="articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a>
    </li>
    <li>
      <a href="articles/influenza-mrt-pcr.html">
        <span class="fa fa-star"></span>

        Influenza A Virus MRT-PCR
      </a>
    </li>
    <li>
      <a href="articles/sars-cov-2-rt-pcr.html">
        <span class="fa fa-star"></span>

        SARS-COV-2 Virus RT-PCR
      </a>
    </li>
    <li>
      <a href="articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a>
    </li>
    <li>
      <a href="articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a>
    </li>
    <li class="dropdown-header">Illumina Library Preparation and Sequencing</li>
    <li>
      <a href="articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a>
    </li>
    <li>
      <a href="articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a>
    </li>
    <li>
      <a href="articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a>
    </li>
    <li>
      <a href="articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a>
    </li>
    <li class="dropdown-header">Oxford Nanopore Library Preparation and Sequencing</li>
    <li>
      <a href="articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html">
        <span class="fa fa-star"></span>

        Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons
      </a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Help

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="articles/troubleshooting.html">Troubleshooting</a>
    </li>
    <li>
      <a href="articles/sequence-qc.html">QC Standards</a>
    </li>
    <li>
      <a href="articles/development-tips.html">Development Tips</a>
    </li>
    <li>
      <a href="articles/nanopore-frameshift-correction.html">Nanopore Frameshift Correction</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Legacy

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/getting-started.html">Retired PC Getting Started</a>
    </li>
    <li>
      <a href="articles/mac-getting-started.html">Retired MAC Getting Started</a>
    </li>
    <li>
      <a href="articles/mira-installation.html">Retired MIRA installation</a>
    </li>
    <li>
      <a href="articles/data-transfer.html">Retired Data Transfer with Mk1C</a>
    </li>
    <li>
      <a href="articles/running-mira.html">Retired Running Genome Assembly -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="articles/running-mira-illumina.html">Retired Running Genome Assembly -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="articles/training-post-mira.html">Retired Post MIRA sequence analysis with Nextclade</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="news/index.html">Change Log</a>
</li>
<li>
  <a href="mailto:idseqsupport@cdc.gov">
    <span class="fa fa-envelope"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA/issues" class="external-link">
    <span class="fa fa-comments"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="id_-mira-interactive-dashboard-for-influenza-sars-cov-2-and-rsv-assembly-and-curation">
<img src="reference/figures/logo.png" align="left" width="190"> MIRA: Interactive Dashboard for Influenza, SARS-CoV-2 and RSV Assembly and Curation<a class="anchor" aria-label="anchor" href="#id_-mira-interactive-dashboard-for-influenza-sars-cov-2-and-rsv-assembly-and-curation"></a>
</h1></div>
<p><img src="https://github.com/cdcgov/mira/workflows/pkgdown/badge.svg" alt="pkgdown"><img src="https://img.shields.io/docker/pulls/cdcgov/mira" alt="Docker pulls"><img src="https://img.shields.io/docker/image-size/cdcgov/mira" alt="Docker image size"><img src="https://img.shields.io/github/last-commit/cdcgov/mira" alt="GitHub last commit"></p>
<p><br></p>
<p><strong>General disclaimer</strong> This repository was created for use by CDC programs to collaborate on public health related projects in support of the <a href="https://www.cdc.gov/about/divisions-offices/index.html" class="external-link">CDC mission</a>. GitHub is not hosted by the CDC, but is a third party website used by CDC and its partners to share information and collaborate on software. CDC use of GitHub does not imply an endorsement of any one particular service, product, or enterprise.</p>
<hr>
<p><br></p>
</div>
<div class="section level1">
<h1 id="run-mira-today">Run MIRA Today!<a class="anchor" aria-label="anchor" href="#run-mira-today"></a>
</h1>
<p><img src="reference/figures/mira-tutorial.gif" alt="Running MIRA" width="800" height="700"></p>
</div>
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p><strong>MIRA</strong> is a bioinformatics pipeline that assembles Influenza genomes, SARS-CoV-2 genomes, the SARS-CoV-2 spike-gene and RSV genomes when given the raw fastq files and a samplesheet. MIRA can analyze reads from both Illumina and OxFord Nanopore sequencing machines. Coming soon, it will allow for upload via FTP to NCBI’s databases <strong>Genbank</strong>, <strong>BioSample</strong>, and <strong>SRA</strong>, as well as <strong>GISAID</strong>.</p>
<p><img src="reference/figures/MIRA_flowchart_pptx.png"></p>
<hr>
<div class="section level2">
<h2 id="which-mira-is-right-for-you">Which MIRA is right for you?<a class="anchor" aria-label="anchor" href="#which-mira-is-right-for-you"></a>
</h2>
<div class="section level3">
<h3 id="running-mira-with-docker-desktop">Running MIRA with Docker Desktop<a class="anchor" aria-label="anchor" href="#running-mira-with-docker-desktop"></a>
</h3>
<p>MIRA-GUI is run on a local computer. It is ideal for users who are not familiar with using the command line, but would still like to perform genomic analysis and render quality figures.</p>
<p>Computational Requirements:</p>
<ul>
<li>A minimum of 16GB of memory is required. &gt;=32GB is recommended.</li>
<li>A minimum of 8 CPU cores. 16 are recommended.</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Understanding of fastq files and their file structure.</li>
</ul>
<p>Key features:</p>
<ul>
<li>Easy to use Graphical User Interface (GUI).</li>
<li>No command line needed/</li>
</ul>
<p>To run MIRA with Docker Desktop you can get started <a href="">here</a>.</p>
</div>
<div class="section level3">
<h3 id="running-mira-cli-with-docker">Running MIRA-CLI with Docker<a class="anchor" aria-label="anchor" href="#running-mira-cli-with-docker"></a>
</h3>
<p>MIRA-CLI is ran on a local computer in the command line. It is ideal for users who are familiar with using the command line and would like to perform genomic analysis and render quality figures.</p>
<p>Computational Requirements:</p>
<ul>
<li>A minimum of 16GB of memory is required. &gt;=32GB is recommended.</li>
<li>A minimum of 8 CPU cores. 16 are recommended.</li>
<li><strong>Administrative privileges are required on a Windows operating system <em>to run linux</em>.</strong></li>
<li>A linux/unix (includes Intel-based* MacOS) operating system is required.
<ul>
<li>*This software has not yet been tested on Apple’s M-chip based OS</li>
<li>If you are using a Mac, <a href="mira-cli-mac-getting-started.htmll">go to Mac Getting Started.</a>
</li>
</ul>
</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Familiarity with command line interface.</li>
<li>Understanding of fastq files and their file structure.</li>
</ul>
<p>Key features:</p>
<ul>
<li>HTML reports.</li>
</ul>
<p>To run MIRA-CLI you can get started with our <a href="quick-start-ubuntu.html">Quick Start for Ubuntu instrcutions</a>, <a href="getting-started.html">PC instructions</a> or <a href="mira-cli-mac-getting-started.html">Mac instructions</a>.</p>
</div>
<div class="section level3">
<h3 id="running-mira-nf-with-nextflow">Running MIRA-NF with Nextflow<a class="anchor" aria-label="anchor" href="#running-mira-nf-with-nextflow"></a>
</h3>
<p>MIRA-NF is ran on a high performing computing (HPC) cluster or in a cloud computing platform. It is ideal for users who are familiar with using the command line, familiar with high computing clusters and would like to perform genomic analysis and render quality figures.</p>
<p>Computational Requirements:</p>
<ul>
<li>Access to an HPC or cloud computing platform.</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Familiarity with analyzing data in HPC and cloud computing platforms.</li>
<li>Strong understanding of command line interface.</li>
<li>Understanding of fastq files and their file structure.</li>
</ul>
<p>Key features:</p>
<ul>
<li>Able to run large data sets without the need for subsampling.</li>
<li>HTML reports.</li>
</ul>
</div>
<div class="section level3">
<h3 id="to-run-mira-nf-you-can-get-started-by-cloning-the-mira-nf-repo">To run MIRA-NF you can get started by cloning the <a href="https://github.com/CDCgov/MIRA-NF" class="external-link">MIRA-NF Repo</a>!<a class="anchor" aria-label="anchor" href="#to-run-mira-nf-you-can-get-started-by-cloning-the-mira-nf-repo"></a>
</h3>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="test-your-mira-setup">Test your MIRA Setup<a class="anchor" aria-label="anchor" href="#test-your-mira-setup"></a>
</h2>
<ul>
<li><a href="https://centersfordiseasecontrol.sharefile.com/d-s839d7319e9b04e2baba07b4d328f02c2" class="external-link">Click here to download tiny test data from ONT Influenza genome and SARS-CoV-2-spike - 40Mb</a></li>
<li><a href="https://centersfordiseasecontrol.sharefile.com/d-s3c52c0b25c2243078f506d60bd787c62" class="external-link">Click here for the above data set + full genomes of Influenza and SARS-CoV-2 from Illumina MiSeqs - 1Gb</a></li>
<li>unzip the file and find two folders:
<ol style="list-style-type: decimal">
<li><code>tiny_test_run_flu</code></li>
<li><code>tiny_test_run_sc2</code></li>
</ol>
</li>
<li>move these folders into <code>MIRA_NGS</code>
<ul>
<li>if you cannot find the MIRA_NGS folder in your Linux section of file explorer, look in Linux–&gt;home–&gt;your username</li>
</ul>
</li>
</ul>
<hr>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MIRA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-8591-3954" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6247-5082" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/rchau88" class="external-link">Reina Chau</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3012-1404" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2007-4210" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>, <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>, <a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
