<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MIRA Documentation • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="MIRA Documentation">
<meta name="description" content="MIRA is an interactive dashboard created using Dash, a python framework written on the top of Flask, Plotly.js and React.js. The Dash dashboard allows users to interactively create a metadata and config file for Influenza , SARS-COV-2 and RSV Assembly. Subsequently, it generates files necessary to upload via FTP to NCBI’s databases Genbank, BioSample, and SRA, as well as, GISAID.">
<meta property="og:description" content="MIRA is an interactive dashboard created using Dash, a python framework written on the top of Flask, Plotly.js and React.js. The Dash dashboard allows users to interactively create a metadata and config file for Influenza , SARS-COV-2 and RSV Assembly. Subsequently, it generates files necessary to upload via FTP to NCBI’s databases Genbank, BioSample, and SRA, as well as, GISAID.">
<meta property="og:image" content="https://cdcgov.github.io/MIRA/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">MIRA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-in-docker-desktop" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA in Docker Desktop</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-in-docker-desktop">
<li><a class="dropdown-item" href="articles/mira-dd-getting-started.html">Getting Started and Installation</a></li>
    <li><a class="dropdown-item" href="articles/running-mira-dd-ont.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="articles/running-mira-dd-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="articles/data-transfer.html">Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="articles/upgrading-mira-dd.html">Upgrading MIRA Docker Desktop</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-cli" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA-CLI</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-cli">
<li><a class="dropdown-item" href="articles/quick-start-ubuntu.html">Quick Start in Ubuntu</a></li>
    <li><a class="dropdown-item" href="articles/getting-started.html">PC Getting Started</a></li>
    <li><a class="dropdown-item" href="articles/mira-cli-mac-getting-started.html">MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="articles/mira-cli-installation.html">MIRA installation</a></li>
    <li><a class="dropdown-item" href="articles/running-mira-cli-ont.html">Running Genome Assembly with MIRA-CLI -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="articles/running-mira-cli-illumina.html">Running Genome Assembly with MIRA-CLI -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="articles/upgrading-mira-cli.html">Upgrading MIRA-CLI</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lab-protocols" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Lab Protocols</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-lab-protocols">
<li><h6 class="dropdown-header" data-toc-skip>Sequencing Template Generation</h6></li>
    <li><a class="dropdown-item" href="articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a></li>
    <li><a class="dropdown-item" href="articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a></li>
    <li><a class="dropdown-item" href="articles/influenza-mrt-pcr.html"><span class="fa fa-star"></span> Influenza A Virus MRT-PCR</a></li>
    <li><a class="dropdown-item" href="articles/sars-cov-2-rt-pcr.html"><span class="fa fa-star"></span> SARS-COV-2 Virus RT-PCR</a></li>
    <li><a class="dropdown-item" href="articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a></li>
    <li><a class="dropdown-item" href="articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Illumina Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a></li>
    <li><a class="dropdown-item" href="articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a></li>
    <li><a class="dropdown-item" href="articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a></li>
    <li><a class="dropdown-item" href="articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Oxford Nanopore Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html"><span class="fa fa-star"></span> Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="articles/troubleshooting.html">Troubleshooting</a></li>
    <li><a class="dropdown-item" href="articles/sequence-qc.html">QC Standards</a></li>
    <li><a class="dropdown-item" href="articles/development-tips.html">Development Tips</a></li>
    <li><a class="dropdown-item" href="articles/nanopore-frameshift-correction.html">Nanopore Frameshift Correction</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-legacy" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Legacy</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-legacy">
<li><a class="dropdown-item" href="articles/getting-started.html">Retired PC Getting Started</a></li>
    <li><a class="dropdown-item" href="articles/mac-getting-started.html">Retired MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="articles/mira-installation.html">Retired MIRA installation</a></li>
    <li><a class="dropdown-item" href="articles/data-transfer.html">Retired Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="articles/running-mira.html">Retired Running Genome Assembly -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="articles/running-mira-illumina.html">Retired Running Genome Assembly -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="articles/training-post-mira.html">Retired Post MIRA sequence analysis with Nextclade</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Change Log</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/CDCgov/MIRA/issues"><span class="fa fa-comments"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/CDCgov/MIRA"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="id_-mira-portable-interactive-application-for-high-quality-influenza-sars-cov-2-and-rsv-genome-assembly-annotation-and-curation">
 MIRA: Portable, Interactive Application for High-Quality Influenza, SARS-CoV-2 and RSV Genome Assembly, Annotation, and Curation<a class="anchor" aria-label="anchor" href="#id_-mira-portable-interactive-application-for-high-quality-influenza-sars-cov-2-and-rsv-genome-assembly-annotation-and-curation"></a>
</h1>
</div>
<p><img src="https://github.com/cdcgov/mira/workflows/pkgdown/badge.svg" alt="pkgdown"><img src="https://img.shields.io/docker/pulls/cdcgov/mira" alt="Docker pulls"><img src="https://img.shields.io/docker/image-size/cdcgov/mira" alt="Docker image size"><img src="https://img.shields.io/github/last-commit/cdcgov/mira" alt="GitHub last commit"></p>
<p><br></p>
<p><strong>General disclaimer</strong> This repository was created for use by CDC programs to collaborate on public health related projects in support of the <a href="https://www.cdc.gov/about/divisions-offices/index.html" class="external-link">CDC mission</a>. GitHub is not hosted by the CDC, but is a third party website used by CDC and its partners to share information and collaborate on software. CDC use of GitHub does not imply an endorsement of any one particular service, product, or enterprise.</p>
<hr>
<p><br></p>
</div>
<div class="section level1">
<h1 id="get-started-with-mira-today">Get Started with MIRA Today!<a class="anchor" aria-label="anchor" href="#get-started-with-mira-today"></a>
</h1>
<p><em><a href="./articles/quick-start-ubuntu.html">Quick start on Ubuntu OS</a></em></p>
<p><img src="reference/figures/mira-tutorial.gif" alt="Running MIRA" width="800" height="700"></p>
</div>
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p><strong>MIRA</strong> is a bioinformatics pipeline that assembles Influenza genomes, SARS-CoV-2 genomes, the SARS-CoV-2 spike-gene and RSV genomes when given the raw fastq files and a samplesheet. MIRA can assemble reads from both Illumina and OxFord Nanopore sequencing machines. <em>Coming soon, it will automate sequence submission to NCBI’s <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link"><strong>Genbank</strong></a>, <a href="https://www.ncbi.nlm.nih.gov/biosample/" class="external-link"><strong>BioSample</strong></a>, and <a href="https://www.ncbi.nlm.nih.gov/sra/" class="external-link"><strong>SRA</strong></a>, as well as <a href="https://gisaid.org/" class="external-link"><strong>GISAID</strong></a> using <a href="https://cdcgov.github.io/seqsender/" class="external-link">SeqSender</a>.</em></p>
<p><img src="reference/figures/MIRA_flowchart_pptx.png"></p>
<hr>
<div class="section level2">
<h2 id="which-mira-is-right-for-you">Which MIRA is right for you?<a class="anchor" aria-label="anchor" href="#which-mira-is-right-for-you"></a>
</h2>
<p>All versions of MIRA process raw sequencing reads into consensus genomes using the same pipeline and quality control. All versions create an interactive quality report and multi-fastas ready for public database submission.</p>
<div class="section level3">
<h3 id="running-mira-with-docker-desktop-mira-gui">Running MIRA with Docker Desktop (MIRA-GUI)<a class="anchor" aria-label="anchor" href="#running-mira-with-docker-desktop-mira-gui"></a>
</h3>
<p>MIRA-GUI is the most popular choice for laboratory scientists. It runs on a local computer and is ideal for users who are not comfortable using the command line.</p>
<p>Computational Requirements:</p>
<ul>
<li>A minimum of 16GB of memory is required. &gt;=32GB is recommended.</li>
<li>A minimum of 8 CPU cores. 16 are recommended.</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Ability to install software on a computer.</li>
</ul>
<p>Key features:</p>
<ul>
<li><strong>No command line needed!</strong></li>
<li>Easy to use Graphical User Interface (GUI).</li>
</ul>
<p>To run MIRA with Docker Desktop you can get started <a href="./articles/mira-dd-getting-started.html">here</a>.</p>
</div>
<div class="section level3">
<h3 id="running-mira-on-the-command-line-mira-cli-with-docker">Running MIRA on the command line (MIRA-CLI) with Docker<a class="anchor" aria-label="anchor" href="#running-mira-on-the-command-line-mira-cli-with-docker"></a>
</h3>
<p>MIRA-CLI runs on a local computer from the command line. It is ideal for users who are familiar with running scripts from a terminal and are interested in automating MIRA analyses into local pipelines.</p>
<p>Computational Requirements:</p>
<ul>
<li>A minimum of 16GB of memory is required. &gt;=32GB is recommended.</li>
<li>A minimum of 8 CPU cores. 16 are recommended.</li>
<li><strong>Administrative privileges are required on a Windows operating system <em>to run linux</em>.</strong></li>
<li>A linux/unix (includes MacOS*) operating system is required.
<ul>
<li>This software <strong>does work</strong> on Apple’s M-chip based OS, <em>however we have observed intermittent instability issues.</em>
</li>
<li>If you are using a Mac, <a href="./articles/mira-cli-mac-getting-started.html">go to Mac Getting Started.</a>
</li>
</ul>
</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Ability to navigate your command line and run commands.</li>
</ul>
<p>Key features:</p>
<ul>
<li>Interactive and sharable HTML reports.</li>
</ul>
<p>To run MIRA-CLI you can get started with our <a href="./articles/quick-start-ubuntu.html">Quick Start for Ubuntu instructions</a>, <a href="./articles/getting-started.html">PC instructions</a> or <a href="./articles/mira-cli-mac-getting-started.html">Mac instructions</a>.</p>
</div>
<div class="section level3">
<h3 id="running-mira-nf-with-nextflow">Running MIRA-NF with Nextflow<a class="anchor" aria-label="anchor" href="#running-mira-nf-with-nextflow"></a>
</h3>
<p>MIRA-NF runs on a high performing computing (HPC) cluster or in a cloud computing platform. It is ideal for users who have access to an HPC and want to process a large volume of samples.</p>
<p>Computational Requirements:</p>
<ul>
<li>Access to an HPC or cloud computing platform.</li>
</ul>
<p>Skill Needed:</p>
<ul>
<li>Familiarity with analyzing data in HPC and cloud computing platforms.</li>
<li>Strong understanding of command line interface.</li>
</ul>
<p>Key features:</p>
<ul>
<li>Able to run large data sets <em>without the need for subsampling.</em>
</li>
<li>Interactive and sharable HTML reports.</li>
</ul>
</div>
<div class="section level3">
<h3 id="to-run-mira-nf-you-can-get-started-by-cloning-the-mira-nf-repo">To run MIRA-NF you can get started by cloning the <a href="https://github.com/CDCgov/MIRA-NF" class="external-link">MIRA-NF Repo</a>!<a class="anchor" aria-label="anchor" href="#to-run-mira-nf-you-can-get-started-by-cloning-the-mira-nf-repo"></a>
</h3>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="test-your-mira-setup">Test your MIRA Setup<a class="anchor" aria-label="anchor" href="#test-your-mira-setup"></a>
</h2>
<ul>
<li><a href="https://centersfordiseasecontrol.sharefile.com/d-s839d7319e9b04e2baba07b4d328f02c2" class="external-link">Click here to download tiny test data from ONT Influenza genome and SARS-CoV-2-spike - 40Mb</a></li>
<li><a href="https://centersfordiseasecontrol.sharefile.com/d-s3c52c0b25c2243078f506d60bd787c62" class="external-link">Click here for the above data set + full genomes of Influenza and SARS-CoV-2 from Illumina MiSeqs - 1Gb</a></li>
<li>unzip the file and find two folders:
<ol style="list-style-type: decimal">
<li><code>tiny_test_run_flu</code></li>
<li><code>tiny_test_run_sc2</code></li>
</ol>
</li>
<li>move these folders into <code>MIRA_NGS</code>
<ul>
<li>if you cannot find the MIRA_NGS folder in your Linux section of file explorer, look in Linux–&gt;home–&gt;your username</li>
</ul>
</li>
</ul>
<hr>
</div>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MIRA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a> <br><small class="roles"> Maintainer, author </small> <a href="https://orcid.org/0000-0002-6247-5082" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/rchau88" class="external-link">Reina Chau</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3012-1404" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2007-4210" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a> <br><small class="roles"> Lead, author </small> <a href="https://orcid.org/0000-0002-8591-3954" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="resources">
<h2 data-toc-skip>Resources</h2>
<ul class="list-unstyled">
<li><p><a href="mailto:idseqsupport@cdc.gov"><span class="fa fa-envelope"></span> Contact Us</a></p></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>, <a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a>, <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
