<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIRA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    NGS-Training

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting-started.html">PC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mac-getting-started.html">MAC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mira-installation.html">MIRA installation</a>
    </li>
    <li>
      <a href="../articles/data-transfer.html">Data Transfer with Mk1C</a>
    </li>
    <li>
      <a href="../articles/running-mira.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="../articles/running-mira-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="../articles/training-post-mira.html">Post MIRA sequence analysis with Nextclade</a>
    </li>
    <li>
      <a href="../articles/upgrading-mira.html">Upgrading MIRA</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lab Protocols

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Sequencing Template Generation</li>
    <li>
      <a href="../articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a>
    </li>
    <li>
      <a href="../articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a>
    </li>
    <li>
      <a href="../articles/influenza-mrt-pcr.html">
        <span class="fa fa-star"></span>

        Influenza A Virus MRT-PCR
      </a>
    </li>
    <li>
      <a href="../articles/sars-cov-2-rt-pcr.html">
        <span class="fa fa-star"></span>

        SARS-COV-2 Virus RT-PCR
      </a>
    </li>
    <li>
      <a href="../articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a>
    </li>
    <li>
      <a href="../articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a>
    </li>
    <li class="dropdown-header">Illumina Library Preparation and Sequencing</li>
    <li>
      <a href="../articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a>
    </li>
    <li>
      <a href="../articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a>
    </li>
    <li>
      <a href="../articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a>
    </li>
    <li>
      <a href="../articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a>
    </li>
    <li class="dropdown-header">Oxford Nanopore Library Preparation and Sequencing</li>
    <li>
      <a href="../articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html">
        <span class="fa fa-star"></span>

        Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons
      </a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Help

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/sequence-qc.html">QC Standards</a>
    </li>
    <li>
      <a href="../articles/development-tips.html">Development Tips</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">Change Log</a>
</li>
<li>
  <a href="../mailto:idseqsupport@cdc.gov">
    <span class="fa fa-envelope"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA/issues" class="external-link">
    <span class="fa fa-comments"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/prod/vignettes/nanopore-frameshift-correction.Rmd" class="external-link"><code>vignettes/nanopore-frameshift-correction.Rmd</code></a></small>
      <div class="hidden name"><code>nanopore-frameshift-correction.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="correcting-frameshifts-in-nanopore-mira-data">Correcting Frameshifts in Nanopore MIRA data<a class="anchor" aria-label="anchor" href="#correcting-frameshifts-in-nanopore-mira-data"></a>
</h2>
<p>Oxford Nanopore data is known to have an increased error rate when
calling bases in a homopolymer region.</p>
<p>Ex: if the sequences is AAAAAA, Nanopore might erroneously basecall
AAAAA (deleting one A) which can lead to frameshifts in your final
assembly data.</p>
<p>The MIRA development team is working on adding features that would
auto-correct these homopolymer indels automatically, but in the
meantime, here are instructions and examples for manually fixing
frameshifts in your data.</p>
<p>We will be using JalView to edit our sequences: <a href="https://www.jalview.org/download/" class="external-link uri">https://www.jalview.org/download/</a></p>
<div class="section level3">
<h3 id="h3n2-na-frameshift-correction">H3N2 NA frameshift correction<a class="anchor" aria-label="anchor" href="#h3n2-na-frameshift-correction"></a>
</h3>
<p>At nucleotide position starting 138, H3N2 NA gene has seven Cs in a
row. Nanopore is prone to erroneously deleting one of the seven Cs
(CC-CCCC).</p>
<p>This corresponds to Amino Acid position 44.</p>
<p>The following is an example of frameshift correction for H3N2 samples
with a frameshift at amino acid 44.</p>
<div class="section level5">
<h5 id="how-to-know-if-my-data-has-a-frameshift">How to know if my data has a frameshift?<a class="anchor" aria-label="anchor" href="#how-to-know-if-my-data-has-a-frameshift"></a>
</h5>
<p>There are a couple of ways to know if your data contains a
frameshift.</p>
<p><strong>1. In MIRA output</strong></p>
<p>In the reference variants table for your MIRA output, frameshifts are
represented by a “~” character. You can specifically filter for
instances of “~” by searching in the final column. Here, I searched for
“NA” in my protein column and “~” in my AA Variants column to see 4
samples in my run with the frameshift at NA amino acid position 44.</p>
<p><img src="images%2Fsearch_frameshifts_MIRA.PNG"></p>
<p></p>
<p><strong>2. In Nextclade output</strong></p>
<p>If you loaded all of your NA samples to Nextclade, any sequences with
a frameshift will be flagged in the output. Here is what Nextclade shows
for these 4 NA samples (which I called aa, bb, hh, and mm):</p>
<ul>
<li>Go to <a href="https://clades.nextstrain.org/" class="external-link uri">https://clades.nextstrain.org/</a>
</li>
<li>Make sure you select H3N2 NA as your pathogen</li>
</ul>
<p><img src="images%2Fnextclade_NA_1.PNG"></p>
<p></p>
<ul>
<li>After you select or input your N2 sequences, if the results show an
orange “F” in th QC column, 1 or more in the Gaps column, and a big red
bar in the gene section at the far right, then you have a
frameshift</li>
</ul>
<p><img src="images%2Fnextclade_NA_2.PNG"></p>
<p><strong>3. Fixing the frameshift with JalView</strong></p>
<ul>
<li>Select the download button in the top right corner of Nextclade’s
output, and choose the <strong>nextclade.aligned.fasta</strong>
</li>
</ul>
<p><img src="images%2Fnextclade_NA_3.PNG"></p>
<ul>
<li>Open the downloaded .fasta and scroll to the position with the
frameshift (if you hover over Gaps in Nextclade, it tells us for these
samples, the gap is at nucleotide position 140)</li>
</ul>
<p><img src="images%2Fgaps.PNG"></p>
<ul>
<li>Since we downloaded the aligned .fasta, we should see a “-”
character at position 140 in JalView. Notice all the Cs there? That is
our homopolymer:</li>
</ul>
<p><img src="images%2Fpos_140.PNG"></p>
<ul>
<li><p>TIP: Is your JalView nucleotide size too small and difficult to
read? Click Format –&gt; Font –&gt; Size to make it bigger</p></li>
<li><p>Next, we need to “fill” this gap so that it is no longer a
frameshift. But since we are missing data, we cannot say for certain
that there should be a C here. We just know that there should be a
nucleotide. Therefore, we will fill this gap with an N
character.</p></li>
<li><p>Select the gap by clicking on the “-”, you should see a red box
around it. Right click then choose Selection, Edit, then Edit
Sequence</p></li>
</ul>
<p><img src="images%2Fedit_sequence_jalview.PNG"></p>
<ul>
<li><p>A text box will appear. Replace the “-” with an uppercase “N” and
then click OK</p></li>
<li><p>You have successfully filled a frameshift gap!</p></li>
<li><p>Replace the original sequences in your MIRA multifasta! First,
open the original .fasta in Jalview and select the frameshift samples,
then type the delete key on your keyboard. Then, you can copy and paste
the “fixed” sequence from your nextclade.aligned.fasta JalView window
into your MIRA multifasta Jalview window and save.</p></li>
<li><p>Now, when you run Nextclade on the “fixed” sequences, you will
see 0 gaps and 0 frameshifts:</p></li>
</ul>
<p><img src="images%2Fnextclade_NA_4.PNG"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>, <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
