<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • MIRA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="MIRA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIRA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    NGS-Training
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting-started.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/mira-installation.html">MIRA installation</a>
    </li>
    <li>
      <a href="../articles/running-mira.html">Running Genome Assembly with MIRA</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Help
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/sequence-qc.html">QC Standards</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">Change Log</a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/HEAD/vignettes/running-mira.Rmd"><code>vignettes/running-mira.Rmd</code></a></small>
      <div class="hidden name"><code>running-mira.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="running-mira">Running MIRA<a class="anchor" aria-label="anchor" href="#running-mira"></a>
</h3>
<p><img src="images/irma-spy.jpg"></p>
<p><em>This documentation assumes you have the four required Docker
containers (<strong>IRMA</strong>, <strong>DAIS-Ribosome</strong>,
<strong>spyne</strong>, and <strong>MIRA</strong>) already running. If
not, please start with <a href="setting-up-containers.html">setting up
containers</a> instructions.</em></p>
<hr>
</div>
<div class="section level3">
<h3 id="stage-your-demultiplexed-data">Stage your demultiplexed data<a class="anchor" aria-label="anchor" href="#stage-your-demultiplexed-data"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>In your “FLU_SC2_SEQUENCING” folder, make a folder for your run.
It is very helpful to name your runs with a consistent naming system,
such as “YEAR-MONTH-DATE-FLOWCELL_ID”, ie. 2022-11-13-ABC1234.</p>
<h2 id="do-not-put-spaces-or-slashes-in-you-run-folder-names"><strong><em>DO
NOT PUT SPACES OR SLASHES IN YOU RUN FOLDER NAMES!</em></strong></h2>
<p><img src="images/mira_run_directory_structure.png"></p>
</li>
<li>
<p>Copy the run’s demultiplexed fastqs into the RUN-FOLDER.</p>
<ol style="list-style-type: decimal">
<li>If this run is from an Oxford Nanopore Technologies’ (ONT)
sequencer, copy the “fastq_pass” directory from the sequencer’s output
into the RUN-FOLDER. On a Mk1C instrument, this is found in
/data/RUN_NAME/EXPERIMENT_NAME/NANOPORE-NAME/fastq_pass where RUN_NAME
and EXPERIMENT_NAME are defined by you in MinKNOW when you set up the
sequencing run on the device and the NANOPORE-NAME is created by the
instrument during the run.</li>
</ol>
<p><img src="images/mira_run_directory_structure_fastq.png"></p>
<ol start="2" style="list-style-type: decimal">
<li><del>If this run is from an Illumina instrument, create another
folder inside your RUN-FOLDER called “fastqs” and copy the folder
containing demultiplexed fastqs into this newly created fastqs folder so
that the folder structure is
FLU_SC2_SEQUENCING/RUN-FOLDER/fastqs/demultiplexed-fastqs/sample-reads.fastq.
There will be two fastq files per sample.</del></li>
</ol>
</li>
</ol>
<hr>
</div>
<div class="section level3">
<h3 id="run-genome-assembly-with-mira">Run genome assembly with MIRA<a class="anchor" aria-label="anchor" href="#run-genome-assembly-with-mira"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>Open Docker Desktop and navigate to the <code>Containers</code>
tab on the left sidebar. Assure that the following containers are
running or run each of them:</p>
<ol style="list-style-type: decimal">
<li>irma</li>
<li>dais</li>
<li>spyne</li>
<li>MIRA</li>
</ol>
</li>
<li>
<p>Run MIRA by clicking the blue link “8020:8050”. This will open
MIRA into your default internet browser. You can “bookmark” this site in
your browser.</p>
<p><img src="images/mira_docker_rolleddown.png"></p>
</li>
<li>
<p>Click the <code>REFRESH RUN LISTING</code> button and select your
run from the dropdown box.</p>
<p><img src="images/mira_run_selection.png"></p>
</li>
<li>
<p>Now, enter your <code>Barcode numbers</code>. Usually you will be
running a simple range of barcode numbers, such as “1-24”, but to allow
for flexibilty, barcode numbers can be entered in a list combining both
single numbers and number-ranges, such as “1-3,5,9,12-20”. Now enter
your <code>Sample ID</code>s. These can be copy and pasted from a column
in an excel sheet. Sample names should only be numbers, letters, and
underscores (“_“). <strong>Do not use any other characters such as ”-”
or ”/”.</strong> Next, select your <code>Sample Type</code>. Most of
your sample’s are test samples, so each row defaults to
<code>test</code>, but select <code>- control</code> or
<code>+ control</code> for your <strong>negative</strong> and
<strong>positive</strong> controls respectively. Finally, select the
<code>Barcode Expansion Pack</code> you have used during library
preparation. CDC’s laboratory protocol is currently
using”EXP-NBD196”.</p>
<ul>
<li>when running tiny_test_run_flu, enter barcode numbers
<code>27,37,41</code> and make up sample names</li>
<li>when running tiny_test_run_sc2, enter barcode numbers
<code>2,3,5,8,28</code> and make up sample names</li>
</ul>
<p><img src="images/mira_samplesheet_barcode_entry.png"><img src="images/mira_samplesheet_fill.png"></p>
</li>
<li><p>Click <code>SAVE SAMPLESHEET</code>. <em>If you need to edit the
number of samples/barcodes in your samplesheet after saving, you must
click</em> <code>RESTART SAMPLESHEET FILLING</code> <em>before editing
it.</em> You will now have to restart filling your samplesheet from the
beginning!</p></li>
<li>
<p>Now select the correct data type from the drop down box,
<code>What kind of data is this?</code> and click
<code>START GENOME ASSEMBLY</code>. Now is a good time to go have a
coffee.</p>
<p><img src="images/mira_whatkindofdata.png"></p>
<p>Click on the toggle button <code>Watch IRMA Progress</code>. Assembly
will take some time, how long exactly is dependent on your sample
number, the number of raw sequencing reads, and the power of your
computer. <img src="images/mira_run_status.png"></p>
</li>
<li>
<p>When IRMA has finished, it will say “IRMA is finished” and you
can now click on “DISPLAY IRMA RESULTS”.</p>
<p><img src="images/mira_display_results.png"></p>
</li>
</ol>
<hr>
</div>
<div class="section level3">
<h3 id="review-irma-results">Review IRMA results<a class="anchor" aria-label="anchor" href="#review-irma-results"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<p>Review the distribution of reads assigned to each barcode. The
ideal result would be a similar number of reads assigned to each test
and positive control. However, it is ok to not have similar read numbers
per sample. Samples with a low proportion of reads may indicate higher
Ct of starting material or less performant PCR during library
preparation. What is most important for sequencing assembly is raw count
of reads and their quality.</p>
<p><img src="images/mira_barcode_distribution.png"></p>
</li>
<li>
<p>Review the “Automatic Quality Control Decisions” heatmap. In
addition to IRMA’s built in quality control, MIRA requires a minimum
median coverage of 50x, a minimum coverage of the reference length of
90%, and less than 10 minor variants &gt;=5%. These are marked in yellow
to orange according to the number of these failure types. Samples that
failed to generate any assembly are marked in red. In addition,
premature stop codons are flagged in yellow. CDC does not submit
sequences with premature stop codons, particularly in HA, NA or
SARS-CoV-2 Spike. Outside of those genes, premature stop codons near the
end of the gene may be ok for submission. Hover your mouse over the
figure to see individual results.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<tbody>
<tr class="odd">
<td><img src="images/mira_auto_qc_heatmap_paleyellow.png" alt="Light yellow"></td>
<td><img src="images/mira_auto_qc_heatmap_yellow.png" alt="Yellow"></td>
</tr>
<tr class="even">
<td><img src="images/mira_auto_qc_heatmap_orange.png" alt="Orange"></td>
<td><img src="images/mira_auto_qc_heatmap_red.png" alt="Red"></td>
</tr>
<tr class="odd">
<td>–</td>
<td>–</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Review <code>Irma Summary</code>. This summarizes the above
information, as well as detailed read assignment to specific referencs,
in a table. Columns can be sorted and filtered, ie.
<code>&gt;1000</code>. Click <code>Export</code> to save this as an
Excel file.</p>
<p><img src="images/mira_irma_summary.png"></p>
</li>
<li>
<p>Review genome coverage depth. The heatmap summarizes the mean
coverage per sample per gene. All the images in MIRA can be interacted
with. Click the gray buttons on the top right of the images. The camera
icon will save the image. Clicking on a sample in the heatmap, or
selecting one in the drop down menu will display two plots. On the left
is a “sankey plot” that shows the number of reads assigned to the
barcode, how many of those passed IRMA’s QC, and how many are assigned
to each gene reference. The plot on the right shows the sample’s
complete coverage per gene. Try toggling the
<code>log y --- linear y</code> button. In log space, you likely need to
reset the view by clicking the gray button in the top right that looks
like an X inside of a box. You can also try panning and zooming the
plot.</p>
<p><img src="images/mira_coverage_heatmap.png"><img src="images/mira_coverage.png"><img src="images/mira_coverage_zooming.png"><img src="images/mira_coverage_zoomed.png"></p>
</li>
<li>
<p>Inspect amino acid variants against popular reference
sequences.</p>
<p><img src="images/mira_aa_vars.png"></p>
</li>
<li>
<p>Inspect minor variation at single nucleotides, insertions and
deletions <em>relative to the sample’s generated consensus
sequence.</em> These tables are for your own usage and not necessary to
review in detail. They can be sorted and filtered and exported to excel
sheets for further analyses.</p>
<p><img src="images/mira_minor_vars.png"></p>
</li>
</ol>
<hr>
</div>
<div class="section level3">
<h3 id="save-your-sequences-in-a-fasta-file">Save your sequences in a fasta file<a class="anchor" aria-label="anchor" href="#save-your-sequences-in-a-fasta-file"></a>
</h3>
<ul>
<li>
<p>Export IRMA’s <em>amended consensus</em> nucleotide sequences and
amino acid fasta files. This includes only those passing MIRA’s QC
criteria and are ready for submission to public databases!</p>
<p><img src="images/mira_fasta_dl.png"><img src="images/mira_fastas.png"></p>
</li>
</ul>
<hr>
<hr>
<div class="section level4">
<h4 id="wow-you-are-doing-such-great-work--time-to-share-your-sequences-with-the-world-by-uploading-to-gisaid-andor-ncbi-and-to-start-analyzing-your-own-data">Wow! You are doing such great work. Time to share your sequences
with the world by uploading to GISAID and/or NCBI and to start analyzing
your own data!<a class="anchor" aria-label="anchor" href="#wow-you-are-doing-such-great-work--time-to-share-your-sequences-with-the-world-by-uploading-to-gisaid-andor-ncbi-and-to-start-analyzing-your-own-data"></a>
</h4>
<ul>
<li>GISAID: <a href="https://gisaid.org/" class="external-link">https://gisaid.org/</a>
</li>
<li>NCBI-Genbank: <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link">https://www.ncbi.nlm.nih.gov/genbank/</a>
</li>
<li>NCBI-BLAST: <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi" class="external-link">https://blast.ncbi.nlm.nih.gov/Blast.cgi</a>
</li>
<li>Nextclade: <a href="https://clades.nextstrain.org/" class="external-link">https://clades.nextstrain.org/</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>, <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
