<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title> • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MIRA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-in-docker-desktop" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA in Docker Desktop</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-in-docker-desktop">
<li><a class="dropdown-item" href="../articles/mira-dd-getting-started.html">Getting Started and Installation</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-dd-ont.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-dd-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/data-transfer.html">Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="../articles/upgrading-mira-dd.html">Upgrading MIRA Docker Desktop</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-cli" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA-CLI</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-cli">
<li><a class="dropdown-item" href="../articles/quick-start-ubuntu.html">Quick Start in Ubuntu</a></li>
    <li><a class="dropdown-item" href="../articles/getting-started.html">PC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-cli-mac-getting-started.html">MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-cli-installation.html">MIRA installation</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-cli-ont.html">Running Genome Assembly with MIRA-CLI -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-cli-illumina.html">Running Genome Assembly with MIRA-CLI -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/upgrading-mira-cli.html">Upgrading MIRA-CLI</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lab-protocols" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Lab Protocols</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-lab-protocols">
<li><h6 class="dropdown-header" data-toc-skip>Sequencing Template Generation</h6></li>
    <li><a class="dropdown-item" href="../articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a></li>
    <li><a class="dropdown-item" href="../articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a></li>
    <li><a class="dropdown-item" href="../articles/influenza-mrt-pcr.html"><span class="fa fa-star"></span> Influenza A Virus MRT-PCR</a></li>
    <li><a class="dropdown-item" href="../articles/sars-cov-2-rt-pcr.html"><span class="fa fa-star"></span> SARS-COV-2 Virus RT-PCR</a></li>
    <li><a class="dropdown-item" href="../articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a></li>
    <li><a class="dropdown-item" href="../articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Illumina Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="../articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a></li>
    <li><a class="dropdown-item" href="../articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a></li>
    <li><a class="dropdown-item" href="../articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Oxford Nanopore Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="../articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html"><span class="fa fa-star"></span> Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="../articles/troubleshooting.html">Troubleshooting</a></li>
    <li><a class="dropdown-item" href="../articles/sequence-qc.html">QC Standards</a></li>
    <li><a class="dropdown-item" href="../articles/development-tips.html">Development Tips</a></li>
    <li><a class="dropdown-item" href="../articles/nanopore-frameshift-correction.html">Nanopore Frameshift Correction</a></li>
  </ul>
</li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-legacy" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Legacy</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-legacy">
<li><a class="dropdown-item" href="../articles/getting-started.html">Retired PC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mac-getting-started.html">Retired MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-installation.html">Retired MIRA installation</a></li>
    <li><a class="dropdown-item" href="../articles/data-transfer.html">Retired Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira.html">Retired Running Genome Assembly -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-illumina.html">Retired Running Genome Assembly -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/training-post-mira.html">Retired Post MIRA sequence analysis with Nextclade</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change Log</a></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/CDCgov/MIRA/issues"><span class="fa fa-comments"></span></a></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/CDCgov/MIRA"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/HEAD/vignettes/running-mira.Rmd"><code>vignettes/running-mira.Rmd</code></a></small>
      <div class="d-none name"><code>running-mira.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="running-mira-cli-with-oxford-nanopore-data">Running MIRA-CLI with Oxford Nanopore Data<a class="anchor" aria-label="anchor" href="#running-mira-cli-with-oxford-nanopore-data"></a>
</h2>
<p><img src="images%2Fmira-logo-midjourney_20230526_rmbkgnd.png"></p>
<p><em>This documentation assumes you have the four required Docker
containers (<strong>IRMA</strong>, <strong>DAIS-Ribosome</strong>,
<strong>spyne</strong>, and <strong>MIRA</strong>) already running. If
not, please start with <a href="getting-started.html">getting
started</a> and <a href="mira-installation.html">MIRA install</a>
instructions.</em></p>
<hr>
</div>
<div class="section level2">
<h2 id="stage-your-demultiplexed-data">Stage your demultiplexed data<a class="anchor" aria-label="anchor" href="#stage-your-demultiplexed-data"></a>
</h2>
<ul>
<li>In your “FLU_SC2_SEQUENCING” folder, make a folder for your run. It
is very helpful to name your runs with a consistent naming system, such
as “YEAR-MONTH-DATE-FLOWCELL_ID”, ie. 2022-11-13-ABC1234.</li>
</ul>
<p><strong><em>DO NOT PUT SPACES OR SLASHES IN YOU RUN FOLDER
NAMES!</em></strong></p>
<p><img src="images%2Fmira_run_directory_structure.png"></p>
<ul>
<li>Copy the run’s demultiplexed fastqs into the RUN-FOLDER.
<ul>
<li>Copy the “fastq_pass” directory from the sequencer’s output into the
RUN-FOLDER. On a Mk1C instrument, this is found in
/data/RUN_NAME/EXPERIMENT_NAME/NANOPORE-NAME/fastq_pass where RUN_NAME
and EXPERIMENT_NAME are defined by you in MinKNOW when you set up the
sequencing run on the device and the NANOPORE-NAME is created by the
instrument during the run.</li>
</ul>
</li>
</ul>
<p><img src="images%2Fmira_run_directory_structure_fastq.png"></p>
<p><br><strong>This screenshot shows “Ubuntu-18.04” but many users just
see “Ubuntu”. This can vary, particularly if you happen to have multiple
Ubuntu versions installed. If you have multiple Ubunut distributions
installed, move the fastqs to whichever “Ubuntu” contains
FLU_SC2_SEQUENCING within
home&gt;USERNAME&gt;FLU_SC2_SEQUENCING</strong></p>
<hr>
</div>
<div class="section level2">
<h2 id="run-genome-assembly-with-mira">Run genome assembly with MIRA<a class="anchor" aria-label="anchor" href="#run-genome-assembly-with-mira"></a>
</h2>
<div class="section level3">
<h3 id="start-the-containers">Start the containers<a class="anchor" aria-label="anchor" href="#start-the-containers"></a>
</h3>
<ul>
<li>Open Docker Desktop and navigate to the <code>Containers</code> tab
on the left sidebar. Assure that the following containers are running or
run each of them:
<ol style="list-style-type: decimal">
<li>irma</li>
<li>dais</li>
<li>spyne</li>
<li>MIRA</li>
</ol>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="open-mira">Open MIRA<a class="anchor" aria-label="anchor" href="#open-mira"></a>
</h3>
<ul>
<li>
<p>Run MIRA by clicking the blue link “8020:8050”. This will open
MIRA into your default internet browser. You can “bookmark” this site in
your browser.</p>
<p><img src="images%2Fmira_docker_rolleddown.png"></p>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="select-runs-data-type-and-enter-sample-information">Select runs, data type, and enter sample information<a class="anchor" aria-label="anchor" href="#select-runs-data-type-and-enter-sample-information"></a>
</h3>
<ul>
<li>
<p>Click the <code>REFRESH RUN LISTING</code> button and select your
run from the dropdown box.</p>
<p><img src="images%2Fmira_run_selection.png"></p>
<ul>
<li>Next, click the box that say “What kind of data is this” and select
Flu-ONT, SC2-Whole-Genome-ONT, or SC2-Spike-Only-ONT</li>
</ul>
<p><img src="images%2Fmira_whatkindofdata_ont.PNG"></p>
</li>
<li>
<p>Next, click <code>DOWNLOAD SAMPLESHEET TEMPLATE</code> and open
the resulting Excel spreadsheet</p>
<p><img src="images%2Fmira_samplesheet_barcode_entry.png"></p>
</li>
<li>
<p>Your <code>Barcode numbers</code> column will be inferred from
the fastq_pass folder inside of your run folder.</p>
<ul>
<li>If you do not see a barcode that you expect, this is because that
barcode does not have a fastq to process.</li>
</ul>
</li>
<li>
<p>Enter your sample IDs in the “Sample ID” column</p>
<ul>
<li>These can be copy and pasted from a column in another excel
sheet.</li>
<li>Sample names should only be numbers, letters, dashes, and
underscores (“-” or “_“). <strong>Do not use any other characters such
as “/”.</strong>
</li>
</ul>
</li>
<li>
<p>Next, select your <code>Sample Type</code>. Most of your sample’s
are test samples, so each row defaults to <code>test</code>, but select
<code>- control</code> or <code>+ control</code> for your
<strong>negative</strong> and <strong>positive</strong> controls
respectively.</p>
<p><img src="images%2Fmira_samplesheet_fill.png"></p>
</li>
<li><p>Save your samplesheet with changes in Excel</p></li>
<li>
<p>Back in MIRA, Click the “Drag and Drop your Samplesheet” to
select your filled-in samplesheet.</p>
<ul>
<li>A table in the application should now appear with your samples!</li>
</ul>
<p><img src="images%2Fmira_dash_table.PNG"></p>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="start-genome-assembly">Start Genome Assembly<a class="anchor" aria-label="anchor" href="#start-genome-assembly"></a>
</h3>
<ul>
<li>
<p>Now select <code>UNLOCK ASSEMBLY BUTTON</code> and click
<code>START GENOME ASSEMBLY</code>. Now is a good time to go have a
coffee.</p>
<ul>
<li>Click on the toggle button <code>Watch IRMA Progress</code>.
Assembly will take some time, how long exactly is dependent on your
sample number, the number of raw sequencing reads, and the power of your
computer. <img src="images%2Fmira_run_status.png">
</li>
<li>as IRMA completes assembly on each sample, they will be listed at
the top of the progress indicator</li>
</ul>
</li>
<li>
<p>When IRMA has finished, it will say “IRMA is finished” and you
can now click on “DISPLAY IRMA RESULTS”.</p>
<p><img src="images%2Fmira_display_results.png"></p>
</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="review-irma-results">Review IRMA results<a class="anchor" aria-label="anchor" href="#review-irma-results"></a>
</h2>
<ol style="list-style-type: decimal">
<li>
<p>Review the distribution of reads assigned to each barcode. The
ideal result would be a similar number of reads assigned to each test
and positive control. However, it is ok to not have similar read numbers
per sample. Samples with a low proportion of reads may indicate higher
Ct of starting material or less performant PCR during library
preparation. What is most important for sequencing assembly is raw count
of reads and their quality.</p>
<p><img src="images%2Fmira_barcode_distribution.png"></p>
</li>
<li>
<p>Review the “Automatic Quality Control Decisions” heatmap. In
addition to IRMA’s built in quality control, MIRA requires a minimum
median coverage of 50x, a minimum coverage of the reference length of
90%, and less than 10 minor variants &gt;=5%. These are marked in yellow
to orange according to the number of these failure types. Samples that
failed to generate any assembly are marked in red. In addition,
premature stop codons are flagged in yellow. CDC does not submit
sequences with premature stop codons, particularly in HA, NA or
SARS-CoV-2 Spike. Outside of those genes, premature stop codons near the
end of the gene may be ok for submission. Hover your mouse over the
figure to see individual results.</p>
<table class="table">
<colgroup>
<col width="50%">
<col width="50%">
</colgroup>
<tbody>
<tr class="odd">
<td><img src="images%2Fmira_auto_qc_heatmap_paleyellow.png" alt="Light yellow"></td>
<td><img src="images%2Fmira_auto_qc_heatmap_yellow.png" alt="Yellow"></td>
</tr>
<tr class="even">
<td><img src="images%2Fmira_auto_qc_heatmap_orange.png" alt="Orange"></td>
<td><img src="images%2Fmira_auto_qc_heatmap_red.png" alt="Red"></td>
</tr>
<tr class="odd">
<td>–</td>
<td>–</td>
</tr>
</tbody>
</table>
</li>
<li>
<p>Review <code>Irma Summary</code>. This summarizes the above
information, as well as detailed read assignment to specific referencs,
in a table. Columns can be sorted and filtered, ie.
<code>&gt;1000</code>. Click <code>Export</code> to save this as an
Excel file.</p>
<p><img src="images%2Fmira_irma_summary.png"></p>
</li>
<li>
<p>Review genome coverage depth. The heatmap summarizes the mean
coverage per sample per gene. All the images in MIRA can be interacted
with. Click the gray buttons on the top right of the images. The camera
icon will save the image. Clicking on a sample in the heatmap, or
selecting one in the drop down menu will display two plots. On the left
is a “sankey plot” that shows the number of reads assigned to the
barcode, how many of those passed IRMA’s QC, and how many are assigned
to each gene reference. The plot on the right shows the sample’s
complete coverage per gene. Try toggling the
<code>log y --- linear y</code> button. In log space, you likely need to
reset the view by clicking the gray button in the top right that looks
like an X inside of a box. You can also try panning and zooming the
plot.</p>
<p><img src="images%2Fmira_coverage_heatmap.png"><img src="images%2Fmira_coverage.png"><img src="images%2Fmira_coverage_zooming.png"><img src="images%2Fmira_coverage_zoomed.png"></p>
</li>
<li>
<p>Inspect amino acid variants against popular reference
sequences.</p>
<p><img src="images%2Fmira_aa_vars.png"></p>
</li>
</ol>
<p>Here, you may see some mutations that do not match Amino Acid
letters.</p>
<div class="section level3">
<h3 id="special-translated-characters">Special translated characters<a class="anchor" aria-label="anchor" href="#special-translated-characters"></a>
</h3>
<p>Translation produces standard amino acid codes with the two
non-standard exceptions (. and ~) listed below. The translation engine
stops when it encounters a stop codon, but corrects itself to continue
in-frame when a frameshift is encountered.</p>
<style type="text/css">
th {
  font-weight: bold;
}

th, td {
  border: 1px solid black
}
</style>
<table class="table">
<tr>
<th>
Character
</th>
<th>
Interpretation
</th>
</tr>
<tr>
<td>
.
</td>
<td>
Missing alignment data
</td>
</tr>
<tr>
<td>
-
</td>
<td>
Gap in alignment (ex: deletion)
</td>
</tr>
<tr>
<td>
~
</td>
<td>
Partial codon (ex: frameshift)
</td>
</tr>
<tr>
<td>
X
</td>
<td>
Ambiguous codon translation
</td>
</tr>
</table>
<p><br></p>
<p>If you have frameshifts in your data, <a href="./nanopore-frameshift-correction.Rmd">click here</a> for
instructions on how to fix them</p>
<ol start="6" style="list-style-type: decimal">
<li>
<p>Inspect minor variation at single nucleotides, insertions and
deletions <em>relative to the sample’s generated consensus
sequence.</em> These tables are for your own usage and not necessary to
review in detail. They can be sorted and filtered and exported to excel
sheets for further analyses.</p>
<p><img src="images%2Fmira_minor_vars.png"></p>
<ul>
<li>Minor variants greater than 5% frequency are shown</li>
<li>Indels greater than 20% frequency are shown</li>
</ul>
</li>
</ol>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="save-your-sequences-in-a-fasta-file">Save your sequences in a fasta file<a class="anchor" aria-label="anchor" href="#save-your-sequences-in-a-fasta-file"></a>
</h2>
<ul>
<li>
<p>Export IRMA’s <em>amended consensus</em> nucleotide sequences and
amino acid fasta files. This includes only those passing MIRA’s QC
criteria and are ready for submission to public databases!</p>
<p><img src="images%2Fmira_fasta_dl.png"></p>
<ul>
<li><p>The <code>DOWNLOAD FASTAS</code> button will only return passing
samples. To investigate failed sequences, click the
<code>DOWNLOAD FAILED FASTAS</code> button. Do not submit failed samples
to public repositories!</p></li>
<li><p>You can open and view these Amino Acid and Nucleotide fastas in
any sequence viewing application <img src="images%2Fmira_fastas.png"></p></li>
</ul>
</li>
</ul>
<hr>
<hr>
<div class="section level3">
<h3 id="wow-you-are-doing-such-great-work--time-to-share-your-sequences-with-the-world-by-uploading-to-gisaid-andor-ncbi-and-to-start-analyzing-your-own-data">Wow! You are doing such great work. Time to share your sequences
with the world by uploading to GISAID and/or NCBI and to start analyzing
your own data!<a class="anchor" aria-label="anchor" href="#wow-you-are-doing-such-great-work--time-to-share-your-sequences-with-the-world-by-uploading-to-gisaid-andor-ncbi-and-to-start-analyzing-your-own-data"></a>
</h3>
<ul>
<li>GISAID: <a href="https://gisaid.org/" class="external-link">https://gisaid.org/</a>
</li>
<li>NCBI-Genbank: <a href="https://www.ncbi.nlm.nih.gov/genbank/" class="external-link">https://www.ncbi.nlm.nih.gov/genbank/</a>
</li>
<li>NCBI-BLAST: <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi" class="external-link">https://blast.ncbi.nlm.nih.gov/Blast.cgi</a>
</li>
<li>Nextclade: <a href="https://clades.nextstrain.org/" class="external-link">https://clades.nextstrain.org/</a>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>, <a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a>, <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
