<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> â€¢ MIRA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<meta property="og:description" content="MIRA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIRA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    NGS-Training
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting-started.html">PC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mac-getting-started.html">MAC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mira-installation.html">MIRA installation</a>
    </li>
    <li>
      <a href="../articles/data-transfer.html">Data Transfer with Mk1C</a>
    </li>
    <li>
      <a href="../articles/running-mira.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="../articles/running-mira-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="../articles/upgrading-mira.html">Upgrading MIRA</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lab Protocols
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/influenza-mrt-pcr.html">Influenza A MRT-PCR</a>
    </li>
    <li>
      <a href="../articles/sars-cov-2-rt-pcr.html">SARS-COV-2 RT-PCR</a>
    </li>
    <li>
      <a href="../articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html">Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Help
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/sequence-qc.html">QC Standards</a>
    </li>
    <li>
      <a href="../articles/development-tips.html">Development Tips</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">Change Log</a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/HEAD/vignettes/sequence-qc.Rmd"><code>vignettes/sequence-qc.Rmd</code></a></small>
      <div class="hidden name"><code>sequence-qc.Rmd</code></div>

    </div>

    
    
<p>Assembled genomes undergo two kinds of quality control. First, at the
IRMA level, reads are filtered according to length and median quality.
During the many rounds of reference editing and refinement, indels and
variants must meet additional thresholds to be incorporated. When
consensus is called, base frequency, quality, and depth are considered
before finalizing the sequence.</p>
<p>Second, consensus sequences undergo additional genome-level QC to
ensure completeness and suitable confidence.</p>
<div class="section level2">
<h2 id="irma-qc">IRMA QC<a class="anchor" aria-label="anchor" href="#irma-qc"></a>
</h2>
<div class="section level3">
<h3 id="nanopore-qc-thresholds">Nanopore QC thresholds<a class="anchor" aria-label="anchor" href="#nanopore-qc-thresholds"></a>
</h3>
<div class="section level4">
<h4 id="flu">Flu<a class="anchor" aria-label="anchor" href="#flu"></a>
</h4>
<ul>
<li>Read length &gt;= 150</li>
<li>Insertion frequency &gt;= 75%</li>
<li>Deletion frequency &gt;= 75%</li>
<li>Minimum coverage to call base &gt;= 50X depth</li>
<li>Average base quality to call &gt;= 10</li>
<li>Mixed <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC</a> base when
Minor Allele Frequency &gt;= 20%</li>
</ul>
</div>
<div class="section level4">
<h4 id="spike-sars-cov-2">Spike SARS-CoV-2<a class="anchor" aria-label="anchor" href="#spike-sars-cov-2"></a>
</h4>
<ul>
<li>Read length &gt;= 125</li>
<li>Insertion frequency &gt;= 50%</li>
<li>Minimum insertion coverage to call base &gt;= 15X depth</li>
<li>Deletion frequency &gt;= 50%</li>
<li>Minimum deletion coverage to call deletion &gt;= 10X depth</li>
<li>Minimum coverage to call base &gt;= 50X depth</li>
<li>Average base quality to call &gt;= 10</li>
<li>Mixed <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC</a> base when
Minor Allele Frequency &gt;= 20%</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="illumina-qc-thresholds">Illumina QC thresholds<a class="anchor" aria-label="anchor" href="#illumina-qc-thresholds"></a>
</h3>
<div class="section level4">
<h4 id="flu-1">Flu<a class="anchor" aria-label="anchor" href="#flu-1"></a>
</h4>
<ul>
<li>Read length &gt;= 125</li>
<li>Read median quality score &gt;= 30</li>
<li>Insertion frequency &gt;= 25%</li>
<li>Deletion frequency &gt;= 60%</li>
<li>Mixed <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC</a> base when
Minor Allele Frequency &gt;= 20%</li>
</ul>
</div>
<div class="section level4">
<h4 id="whole-genome-sars-cov-2">Whole Genome SARS-CoV-2<a class="anchor" aria-label="anchor" href="#whole-genome-sars-cov-2"></a>
</h4>
<ul>
<li>Read length &gt;= 80</li>
<li>Read median quality score &gt;= 27</li>
<li>Insertion frequency &gt;= 55%</li>
<li>Minimum insertion coverage to call base &gt;= 15X depth (30X
paired)</li>
<li>Deletion frequency &gt;= 55%</li>
<li>Minimum deletion coverage to call deletion &gt;= 10X depth (30X
paired)</li>
<li>Minimum depth to call base &gt;= 30X depth</li>
<li>Average base quality to call &gt;= 15</li>
<li>Mixed <a href="https://www.bioinformatics.org/sms/iupac.html" class="external-link">IUPAC</a> base when
Minor Allele Frequency &gt;= 20%</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="post-assembly-qc-curation">Post-Assembly QC Curation<a class="anchor" aria-label="anchor" href="#post-assembly-qc-curation"></a>
</h2>
<ul>
<li>Minimum median coverage of 50x</li>
<li>For SARS-CoV-2: Spike gene median coverage of 100X</li>
<li>Minimum coverage of the reference length of 90%</li>
<li>For SARS-CoV-2: Spike gene minimum coverage length of 90%</li>
<li>Less than 10 minor variants &gt;=5% for each Flu Segment</li>
<li>Less than 20 minor variants &gt;=5% for SARS-CoV-2 whole-genome</li>
<li>No premature stop codons in Spike (SC2), HA (Flu), or NA (Flu)
Illumina sequencing</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>, <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
