<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title> • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MIRA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-in-docker-desktop" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA in Docker Desktop</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-in-docker-desktop">
<li><a class="dropdown-item" href="../articles/mira-dd-getting-started.html">Getting Started and Installation</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-dd-ont.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-dd-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/data-transfer.html">Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="../articles/upgrading-mira-dd.html">Upgrading MIRA Docker Desktop</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-mira-cli" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">MIRA-CLI</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-mira-cli">
<li><a class="dropdown-item" href="../articles/quick-start-ubuntu.html">Quick Start in Ubuntu</a></li>
    <li><a class="dropdown-item" href="../articles/getting-started.html">PC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-cli-mac-getting-started.html">MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-cli-installation.html">MIRA installation</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-cli-ont.html">Running Genome Assembly with MIRA-CLI -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-cli-illumina.html">Running Genome Assembly with MIRA-CLI -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/upgrading-mira-cli.html">Upgrading MIRA-CLI</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lab-protocols" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Lab Protocols</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-lab-protocols">
<li><h6 class="dropdown-header" data-toc-skip>Sequencing Template Generation</h6></li>
    <li><a class="dropdown-item" href="../articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a></li>
    <li><a class="dropdown-item" href="../articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a></li>
    <li><a class="dropdown-item" href="../articles/influenza-mrt-pcr.html"><span class="fa fa-star"></span> Influenza A Virus MRT-PCR</a></li>
    <li><a class="dropdown-item" href="../articles/sars-cov-2-rt-pcr.html"><span class="fa fa-star"></span> SARS-COV-2 Virus RT-PCR</a></li>
    <li><a class="dropdown-item" href="../articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a></li>
    <li><a class="dropdown-item" href="../articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Illumina Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="../articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a></li>
    <li><a class="dropdown-item" href="../articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a></li>
    <li><a class="dropdown-item" href="../articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a></li>
    <li><h6 class="dropdown-header" data-toc-skip>Oxford Nanopore Library Preparation and Sequencing</h6></li>
    <li><a class="dropdown-item" href="../articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html"><span class="fa fa-star"></span> Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/FAQs.html">FAQs</a></li>
    <li><a class="dropdown-item" href="../articles/troubleshooting.html">Troubleshooting</a></li>
    <li><a class="dropdown-item" href="../articles/sequence-qc.html">QC Standards</a></li>
    <li><a class="dropdown-item" href="../articles/development-tips.html">Development Tips</a></li>
    <li><a class="dropdown-item" href="../articles/nanopore-frameshift-correction.html">Nanopore Frameshift Correction</a></li>
  </ul>
</li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-legacy" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Legacy</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-legacy">
<li><a class="dropdown-item" href="../articles/getting-started.html">Retired PC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mac-getting-started.html">Retired MAC Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/mira-installation.html">Retired MIRA installation</a></li>
    <li><a class="dropdown-item" href="../articles/data-transfer.html">Retired Data Transfer with Mk1C</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira.html">Retired Running Genome Assembly -- NANOPORE DATA</a></li>
    <li><a class="dropdown-item" href="../articles/running-mira-illumina.html">Retired Running Genome Assembly -- ILLUMINA DATA</a></li>
    <li><a class="dropdown-item" href="../articles/training-post-mira.html">Retired Post MIRA sequence analysis with Nextclade</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change Log</a></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/CDCgov/MIRA/issues"><span class="fa fa-comments"></span></a></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/CDCgov/MIRA"><span class="fa fa-github"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/HEAD/vignettes/training-post-mira.Rmd"><code>vignettes/training-post-mira.Rmd</code></a></small>
      <div class="d-none name"><code>training-post-mira.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="post-mira-consensus-sequence-analysis-practical">Post-MIRA consensus sequence analysis practical<a class="anchor" aria-label="anchor" href="#post-mira-consensus-sequence-analysis-practical"></a>
</h2>
<a href="./training-answers.html"><em>Answers here</em></a>
<hr>
<div class="section level3">
<h3 id="influenza">Influenza<a class="anchor" aria-label="anchor" href="#influenza"></a>
</h3>
<p>Using MIRA-passed, influenza consensus sequences, you will determine
their clades and how different they are from the vaccine strains <a href="https://www.who.int/teams/global-influenza-programme/vaccines/who-recommendations" class="external-link">recommended
by the World Health Organization for the nothern hemisphere 2022/23
season</a>.</p>
<div class="section level4">
<h4 id="goals">Goals<a class="anchor" aria-label="anchor" href="#goals"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Assess the quality of your consensus sequences with MIRA.</li>
<li>Identify which <a href="https://clades.nextstrain.org/" class="external-link">nextclade</a> module to use to
analyze consensus sequences.</li>
<li>Gather the required consensus sequences into a file and upload to
nextclade.</li>
<li>Identify the HA and NA subtypes for each consensus sequence.</li>
<li>Identify number of variants in consensus sequences.</li>
<li>View where sequences are placed on nextstrain reference tree.</li>
<li>Manually edit frameshifts with <a href="https://www.jalview.org/" class="external-link">Jalview</a>
</li>
</ol>
</div>
<div class="section level4">
<h4 id="practical">Practical<a class="anchor" aria-label="anchor" href="#practical"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<a href="https://centersfordiseasecontrol.sharefile.com/d-s4d3bbf8ecfeb4df18ed546c7af8286eb" class="external-link">Click
here to download testdata</a>. <em>You will be required to put in your
name and email.</em>
</li>
<li>Unzip the file into the folder that MIRA accessess runs (ie.
FLU_SC2_SEQUENCING).</li>
<li>Open MIRA and load the run “2023_Paris_Flu_training”
<ul>
<li>Which sample had the fewest reads? How many reads did it have?</li>
</ul>
</li>
<li>Does the negative control pass?</li>
<li>What are some common reasons that segments do not pass quality
control?</li>
<li>What HA and NA subtypes are present?</li>
<li>Take a look at the Reference Variants of the HA segment in sample
“aa”.
<ul>
<li>What do the X’s mean and why are so many of the amino acids marked
this way in sample “aa”?</li>
</ul>
</li>
<li>Which samples and segments have minor variants greater than 20%?
What does this mean for the amended consensus sequences?</li>
<li>Download the fasta sequences and open in JalView or your preferred
sequence viewer.</li>
<li>In your internet browser, navigate to <a href="https://clades.nextstrain.org/" class="external-link">Nextclade</a>.</li>
<li>What Flu A subtypes have modules available in Nextclade?</li>
<li>Which pathogen module uses the WHO recommended vaccine virus for the
H1 2022/23 seasonal component as a reference?</li>
<li>Select that module and copy the HA sequence from the H1N1 sample
from your sequence viewer and paste into the sequence input box and run
Nextclade.</li>
<li>What QC metric is flagged as being a concern for this sequence? What
position in the sequence is the problem?</li>
<li>Download the aligned fasta sequence from nextclade to manually edit
the sequence and fix the frameshift. Given this is a single base
deletion that causes a frameshift, we can confidently say this is an
error in sequencing. Insertion and deletion errors are unfortunately
common from nanopore sequencing, particularly at homopolymer sequences.
<img src="images%2Fpost-mira-training-answers-nextclade-dl-alignment.png">
</li>
</ol>
<p><img src="images%2Fpost-mira-training-answers-jalview-editseq.png"></p>
<p><img src="images%2Fpost-mira-training-answers-jalview-editseq-n.png"></p>
<p>After copying the edited sequences, open the amended_consensus.fasta
file from MIRA, delete the offending sequence and add in the corrected
sequence from your clipboard and save this as your updated fasta file to
be submitted to GISAID/NCBI. <img src="images%2Fpost-mira-training-answers-jalview-editseq-paste.png"> 16.
Repeat the nextclade analysis with the corrected sequence. How many
mutations does it have versus the reference? 17. View where the sequence
is placed onto a nextclade reference tree by clicking in the top right
corner. <img src="images%2Fpost-mira-training-answers-nextclade-corrected-h1ha-clicktree.png"></p>
<p><img src="images%2Fpost-mira-training-answers-nextclade-corrected-h1ha-tree.png"></p>
</div>
<div class="section level4">
<h4 id="h1n2">H1N2<a class="anchor" aria-label="anchor" href="#h1n2"></a>
</h4>
<p>Nextclade only supports seasonal human influenza H1N1 and H3N2 at
this time. To understand what the HA is most like from the H1N2
sequence, we can use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;BLAST_SPEC=GeoBlast&amp;PAGE_TYPE=BlastSearch" class="external-link">NCBI
BLAST</a>.</p>
<ol start="18" style="list-style-type: decimal">
<li>Copy the HA from the H1N2 sequence into the BLAST Query box, leave
the settings at the default, and click BLAST at the bottom. <img src="images%2Fpost-mira-training-answers-blast-input.png">
</li>
<li>What host is the likely origin of this sequence?</li>
</ol>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="sars-cov-2">SARS-CoV-2<a class="anchor" aria-label="anchor" href="#sars-cov-2"></a>
</h3>
<p>Using MIRA-passed, SARS-CoV-2 spike consensus sequences, determine
their clades and how different they are from Wuhan-1.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>, <a href="https://github.com/mandysulli" class="external-link">Amanda Sullivan</a>, <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
