<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title> • MIRA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MIRA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    NGS-Training

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting-started.html">PC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mac-getting-started.html">MAC Getting Started</a>
    </li>
    <li>
      <a href="../articles/mira-installation.html">MIRA installation</a>
    </li>
    <li>
      <a href="../articles/data-transfer.html">Data Transfer with Mk1C</a>
    </li>
    <li>
      <a href="../articles/running-mira.html">Running Genome Assembly with MIRA -- NANOPORE DATA</a>
    </li>
    <li>
      <a href="../articles/running-mira-illumina.html">Running Genome Assembly with MIRA -- ILLUMINA DATA</a>
    </li>
    <li>
      <a href="../articles/training-post-mira.html">Post MIRA sequence analysis with Nextclade</a>
    </li>
    <li>
      <a href="../articles/upgrading-mira.html">Upgrading MIRA</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Lab Protocols

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Sequencing Template Generation</li>
    <li>
      <a href="../articles/automated-rna-extraction.html">Automated RNA Extraction using QIAcube HT</a>
    </li>
    <li>
      <a href="../articles/influenza-a-and-b-mrt-pcr.html">Multi-segment Reverse Transcription-PCR (MRT-PCR) of Influenza A and B Viruses</a>
    </li>
    <li>
      <a href="../articles/influenza-mrt-pcr.html">
        <span class="fa fa-star"></span>

        Influenza A Virus MRT-PCR
      </a>
    </li>
    <li>
      <a href="../articles/sars-cov-2-rt-pcr.html">
        <span class="fa fa-star"></span>

        SARS-COV-2 Virus RT-PCR
      </a>
    </li>
    <li>
      <a href="../articles/mrt-pcr-purification.html">MRT-PCR Purification using Exonuclease</a>
    </li>
    <li>
      <a href="../articles/QIAxcel-sample-preparation-and-quality-control.html">QIAxcel Sample Preparation and Quality Control Procedure</a>
    </li>
    <li class="dropdown-header">Illumina Library Preparation and Sequencing</li>
    <li>
      <a href="../articles/manual-illumina-dna-prep-library.html">Manual Procedure Illumina DNA Library Preparation</a>
    </li>
    <li>
      <a href="../articles/dna-quantification.html">DNA Quantification with Qubit ds DNA High Sensitivity Kit</a>
    </li>
    <li>
      <a href="../articles/miseq-next-generation-seq-run.html">Performing a MiSeq Next Generation Sequencing Run</a>
    </li>
    <li>
      <a href="../articles/miseq-library-loading-preparation.html">Preparation of the DNA Libraries for Loading onto the MiSeq</a>
    </li>
    <li class="dropdown-header">Oxford Nanopore Library Preparation and Sequencing</li>
    <li>
      <a href="../articles/influenza-sars-cov-2-s-gene-amplicons-nanopore.html">
        <span class="fa fa-star"></span>

        Library Preparation and Nanopore Sequencing of Influenza and SARS-COV-2 S-gene Amplicons
      </a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Help

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FAQs.html">FAQs</a>
    </li>
    <li>
      <a href="../articles/troubleshooting.html">Troubleshooting</a>
    </li>
    <li>
      <a href="../articles/sequence-qc.html">QC Standards</a>
    </li>
    <li>
      <a href="../articles/development-tips.html">Development Tips</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../news/index.html">Change Log</a>
</li>
<li>
  <a href="../mailto:idseqsupport@cdc.gov">
    <span class="fa fa-envelope"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA/issues" class="external-link">
    <span class="fa fa-comments"></span>

  </a>
</li>
<li>
  <a href="https://github.com/CDCgov/MIRA" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/CDCgov/MIRA/blob/prod/vignettes/training-post-mira.Rmd" class="external-link"><code>vignettes/training-post-mira.Rmd</code></a></small>
      <div class="hidden name"><code>training-post-mira.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="post-mira-consensus-sequence-analysis-practical">Post-MIRA consensus sequence analysis practical<a class="anchor" aria-label="anchor" href="#post-mira-consensus-sequence-analysis-practical"></a>
</h2>
<a href="./training-answers.html"><em>Answers here</em></a>
<hr>
<div class="section level3">
<h3 id="influenza">Influenza<a class="anchor" aria-label="anchor" href="#influenza"></a>
</h3>
<p>Using MIRA-passed, influenza consensus sequences, you will determine
their clades and how different they are from the vaccine strains <a href="https://www.who.int/teams/global-influenza-programme/vaccines/who-recommendations" class="external-link">recommended
by the World Health Organization for the nothern hemisphere 2022/23
season</a>.</p>
<div class="section level4">
<h4 id="goals">Goals<a class="anchor" aria-label="anchor" href="#goals"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Assess the quality of your consensus sequences with MIRA.</li>
<li>Identify which <a href="https://clades.nextstrain.org/" class="external-link">nextclade</a> module to use to
analyze consensus sequences.</li>
<li>Gather the required consensus sequences into a file and upload to
nextclade.</li>
<li>Identify the HA and NA subtypes for each consensus sequence.</li>
<li>Identify number of variants in consensus sequences.</li>
<li>View where sequences are placed on nextstrain reference tree.</li>
<li>Manually edit frameshifts with <a href="https://www.jalview.org/" class="external-link">Jalview</a>
</li>
</ol>
</div>
<div class="section level4">
<h4 id="practical">Practical<a class="anchor" aria-label="anchor" href="#practical"></a>
</h4>
<ol style="list-style-type: decimal">
<li>
<a href="https://centersfordiseasecontrol.sharefile.com/d-s4d3bbf8ecfeb4df18ed546c7af8286eb" class="external-link">Click
here to download testdata</a>. <em>You will be required to put in your
name and email.</em>
</li>
<li>Unzip the file into the folder that MIRA accessess runs (ie.
FLU_SC2_SEQUENCING).</li>
<li>Open MIRA and load the run “2023_Paris_Flu_training”
<ul>
<li>Which sample had the fewest reads? How many reads did it have?</li>
</ul>
</li>
<li>Does the negative control pass?</li>
<li>What are some common reasons that segments do not pass quality
control?</li>
<li>What HA and NA subtypes are present?</li>
<li>Take a look at the Reference Variants of the HA segment in sample
“aa”.
<ul>
<li>What do the X’s mean and why are so many of the amino acids marked
this way in sample “aa”?</li>
</ul>
</li>
<li>Which samples and segments have minor variants greater than 20%?
What does this mean for the amended consensus sequences?</li>
<li>Download the fasta sequences and open in JalView or your preferred
sequence viewer.</li>
<li>In your internet browser, navigate to <a href="https://clades.nextstrain.org/" class="external-link">Nextclade</a>.</li>
<li>What Flu A subtypes have modules available in Nextclade?</li>
<li>Which pathogen module uses the WHO recommended vaccine virus for the
H1 2022/23 seasonal component as a reference?</li>
<li>Select that module and copy the HA sequence from the H1N1 sample
from your sequence viewer and paste into the sequence input box and run
Nextclade.</li>
<li>What QC metric is flagged as being a concern for this sequence? What
position in the sequence is the problem?</li>
<li>Download the aligned fasta sequence from nextclade to manually edit
the sequence and fix the frameshift. Given this is a single base
deletion that causes a frameshift, we can confidently say this is an
error in sequencing. Insertion and deletion errors are unfortunately
common from nanopore sequencing, particularly at homopolymer sequences.
<img src="images%2Fpost-mira-training-answers-nextclade-dl-alignment.png">
</li>
</ol>
<p><img src="images%2Fpost-mira-training-answers-jalview-editseq.png"></p>
<p><img src="images%2Fpost-mira-training-answers-jalview-editseq-n.png"></p>
<p>After copying the edited sequences, open the amended_consensus.fasta
file from MIRA, delete the offending sequence and add in the corrected
sequence from your clipboard and save this as your updated fasta file to
be submitted to GISAID/NCBI. <img src="images%2Fpost-mira-training-answers-jalview-editseq-paste.png"> 16.
Repeat the nextclade analysis with the corrected sequence. How many
mutations does it have versus the reference? 17. View where the sequence
is placed onto a nextclade reference tree by clicking in the top right
corner. <img src="images%2Fpost-mira-training-answers-nextclade-corrected-h1ha-clicktree.png"></p>
<p><img src="images%2Fpost-mira-training-answers-nextclade-corrected-h1ha-tree.png"></p>
</div>
<div class="section level4">
<h4 id="h1n2">H1N2<a class="anchor" aria-label="anchor" href="#h1n2"></a>
</h4>
<p>Nextclade only supports seasonal human influenza H1N1 and H3N2 at
this time. To understand what the HA is most like from the H1N2
sequence, we can use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&amp;BLAST_SPEC=GeoBlast&amp;PAGE_TYPE=BlastSearch" class="external-link">NCBI
BLAST</a>.</p>
<ol start="18" style="list-style-type: decimal">
<li>Copy the HA from the H1N2 sequence into the BLAST Query box, leave
the settings at the default, and click BLAST at the bottom. <img src="images%2Fpost-mira-training-answers-blast-input.png">
</li>
<li>What host is the likely origin of this sequence?</li>
</ol>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="sars-cov-2">SARS-CoV-2<a class="anchor" aria-label="anchor" href="#sars-cov-2"></a>
</h3>
<p>Using MIRA-passed, SARS-CoV-2 spike consensus sequences, determine
their clades and how different they are from Wuhan-1.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/nbx0" class="external-link">Ben Rambo-Martin</a>, <a href="https://github.com/kristinelacek" class="external-link">Kristine Lacek</a>, <a href="https://github.com/rchau88" class="external-link">Reina Chau</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
